examples/ex24.cpp

```diff
@@ -0,0 +1,279 @@
+
+int dim;
+
+int main(int argc, char *argv[])
+{
+   // 1. Parse command-line options.
+   const char *mesh_file = "../data/beam-hex.mesh";
+   int order = 1;
+   bool static_cond = false;
+   bool pa = false;
+   bool visualization = 1;
+
+   OptionsParser args(argc, argv);
+   args.AddOption(&mesh_file, "-m", "--mesh",
+                  "Mesh file to use.");
+   args.AddOption(&order, "-o", "--order",
+                  "Finite element order (polynomial degree).");
+   args.AddOption(&static_cond, "-sc", "--static-condensation", "-no-sc",
+                  "--no-static-condensation", "Enable static condensation.");
+   args.AddOption(&pa, "-pa", "--partial-assembly", "-no-pa",
+                  "--no-partial-assembly", "Enable Partial Assembly.");
+   args.AddOption(&visualization, "-vis", "--visualization", "-no-vis",
+                  "--no-visualization",
+                  "Enable or disable GLVis visualization.");
+
+   args.Parse();
+   if (!args.Good())
+   {
+      args.PrintUsage(cout);
+      return 1;
+   }
+   args.PrintOptions(cout);
+
+   // 2. Read the mesh from the given mesh file. We can handle triangular,
+   //    quadrilateral, tetrahedral, hexahedral, surface and volume meshes with
+   //    the same code.
+   Mesh *mesh = new Mesh(mesh_file, 1, 1);
+   dim = mesh->Dimension();
+   int sdim = mesh->SpaceDimension();
+
+   // 3. Refine the mesh to increase the resolution. In this example we do
+   //    'ref_levels' of uniform refinement. We choose 'ref_levels' to be the
+   //    largest number that gives a final mesh with no more than 50,000
+   //    elements.
+   {
+      int ref_levels = (int)floor(log(50000./mesh->GetNE())/log(2.)/dim);
+      for (int l = 0; l < ref_levels; l++)
+      {
+         mesh->UniformRefinement();
+      }
+   }
+   mesh->ReorientTetMesh();
+
+   // 4. Define a parallel finite element space on the parallel mesh. Here we
+   //    use the Nedelec finite elements of the specified order.
+   FiniteElementCollection *fec = new ND_FECollection(order, dim);
+   FiniteElementCollection *H1fec = new H1_FECollection(order, dim);
+   FiniteElementSpace *fespace = new FiniteElementSpace(mesh, fec);
+   FiniteElementSpace *H1fespace = new FiniteElementSpace(mesh, H1fec);
+
+   int size = fespace->GetTrueVSize();
+   int H1size = H1fespace->GetTrueVSize();
+   cout << "Number of Nedelec finite element unknowns: " << size << endl;
+   cout << "Number of H1 finite element unknowns: " << H1size << endl;
+
+   // 5. Define the solution vector x as a parallel finite element grid function
+   //    corresponding to fespace. Initialize x by projecting the exact
+   //    solution. Note that only values from the boundary edges will be used
+   //    when eliminating the non-homogeneous boundary condition to modify the
+   //    r.h.s. vector b.
+   GridFunction x(fespace);
+   FunctionCoefficient p_coef(p_exact);
+   GridFunction p(H1fespace);
+   p.ProjectCoefficient(p_coef);
```

Comment:
I was looking at the errors on AMR meshes in parallel and that made me realize that after `ProjectCoefficient` the `p` `GridFunction` is not actually fully conforming (the slave values in the `GridFunction` are not interpolated from their masters, they are simply interpolated from the coefficient). To get a fully conforming `p`, one can call:
```c++
   p.SetTrueVector();
   p.SetFromTrueVector();
```
after `ProjectCoefficient`.

Reply:
I think in the parallel version, ex24p, the `p` `GridFunction` has the same "issue" and can be "fixed" in the same way. Otherwise, unlike the serial version, the parallel version seems to work fine with AMR meshes. Note that in the parallel version there are sample runs with AMR meshes but not in the serial version.

Reply:
Thanks @v-dobrev for finding this issue. I had just copied sample runs from another example (probably ex3), discarding ones that did not work (e.g. AMR for ex24, which is why it is in ex24p but not ex24). Supporting AMR tests was not my goal in this PR, but if we can fix this easily then we should. Otherwise, I would be fine with removing all AMR sample runs from ex24p. However, I have a few questions.

1) Adding the two lines of code you suggested in ex24 did not work for me (still large error for the AMR meshes). Does it work for you, or is there something else that needs to be done?
2) Adding those two lines is fine in ex24p, which works with or without the change. Is it really necessary?
3) If these additional lines are necessary in general for setting DOF's correctly on AMR meshes, why are they not in other examples (e.g. ex3)? I see this only in ex10/ex10p.
4) If this change is necessary (in this example and others), should we simplify by doing it somehow in ProjectCoefficient or some other hidden way so the user does not need to know about this issue and remember the fix after every projection? Or should we at least document in the examples that use it?

Reply:
The reason I noticed this is the error printed at the end of the example:
```console
Solution of (E_h,v) = (grad p_h,v) for E_h and v in H(curl): || E_h - grad p ||_{L^2} = 2.03637e-05

Gradient interpolant E_h = grad p_h in H(curl): || E_h - grad p ||_{L^2} = 2.0294e-05

Projection E_h of exact grad p in H(curl): || E_h - grad p ||_{L^2} = 2.0294e-05
```
from the sample run
```console
mpirun -np 4 ex24p -m ../data/amr-quad.mesh -o 2 -no-vis
```
As you can see the first two errors here are different which is unexpected -- for all conforming meshes they are the same. When you add the "fix" in parallel, the first two norms should be the same, I think. By the way, similar procedure should be used when projecting the exact `grad_p` in H(curl), so the third norm will also be different probably.

I agree with you that it makes sense to add these additional projections inside the `ProjectCoefficient` methods.

To make ex24 work I think one needs to also address the issue I pointed out here: https://github.com/mfem/mfem/pull/1204#discussion_r384249696

Reply:
By the way, I do not insist to fix any of these issues right now -- it is up to you if you want to do it. I just wanted to let you know.

Reply:
Thanks @v-dobrev. I fixed it and now the AMR tests are working in serial and parallel. You are right, that the first two norms are always identical now.